src/flextopo.jl

```diff
@@ -0,0 +1,795 @@
+        flextopo.hortonpondingstorage_m[i] = sum(flextopo.hortonpondingstorage[i] .* flextopo.hrufrac[i])
+        flextopo.hortonevap_m[i] = sum(flextopo.hortonevap[i] .* flextopo.hrufrac[i])
+    end
+end
+
+function hortonponding(flextopo::FLEXTOPO, config)
+    k = flextopo.kclass[1]
+    for i = 1:flextopo.n
+
+        #calculate reduction of shmax due to frost
+        facc = min(flextopo.facc[i][k] + flextopo.temperature[i] * flextopo.kmf[i][k], 0.0) #TODO check units kmf, facc
+        ft = min(
+            max(
+                facc / (flextopo.facc1[i][k] - flextopo.facc0[i][k]) -
+                flextopo.facc0[i][k] / (flextopo.facc1[i][k] - flextopo.facc0[i][k]),
+                flextopo.shmin[i][k],
+            ),
+            1.0,
+        )
+        shmax_frost = ft * flextopo.shmax[i][k]
+
+        #add effective precipitation from interception to the horton ponding storage.  
+        hortonpondingstorage = flextopo.hortonpondingstorage[i][k] + flextopo.precipeffective[i][k]
+        # if soil is filled until max capacity, additional water runs of directly
+        directrunoff_h = max(hortonpondingstorage - shmax_frost, 0.0)
+        #update hortonpondingstorage
+        hortonpondingstorage = hortonpondingstorage - directrunoff_h
+        #net water which infiltrates in horton (careful, shmax_frost can decrease from one timestep to another)
+        netin_hortonpond = max(flextopo.precipeffective[i][k] - directrunoff_h, 0)
+
+        #evaporation from the horton ponding storage
+        hortonevap =
+            hortonpondingstorage > shmax_frost * flextopo.lp[i][k] ? min(hortonpondingstorage, flextopo.potsoilevap[i][k]) :
+            min(flextopo.potsoilevap[i][k] * (hortonpondingstorage / (shmax_frost * flextopo.lp[i][k])))
+        #update storage
+        hortonpondingstorage = hortonpondingstorage - hortonevap
+        #update restevap 
+        restevap = max(0.0, flextopo.potsoilevap[i][k] - hortonevap)
+
+        #excess water from beta function of netin_hortonpond (due to rouding hortonpondingstorage could be slightly larger than shmax_frost, make sure it is always less than 1)
+        qhorton_in = netin_hortonpond * (1 - pow((1 - min(hortonpondingstorage / shmax_frost, 1)), flextopo.beta[i][k]))
+        #update storage
+        hortonpondingstorage = hortonpondingstorage - qhorton_in
+
+        #total water out of horton ponding consists of directrunoff and excess water from beta function
+        qhortonpond = directrunoff_h + qhorton_in
+
+        #infiltration to root-zone 
+        # qhortonrootzone = max(0.0, flextopo.fmax[i][k] * exp(-flextopo.fdec[i][k] * (1 - hortonpondingstorage/shmax_frost)))
+        qhortonrootzone =
+            min(hortonpondingstorage / shmax_frost, 1) > 0 ?
+            min(
+                flextopo.fmax[i][k] *
+                exp(-flextopo.fdec[i][k] * (1 - min(hortonpondingstorage / shmax_frost, 1))),
+                hortonpondingstorage,
+            ) : 0
```

Comment:
This probably should be 0.0, to avoid conditionally making it an integer.

Reply:
indeed, I noticed that I had used integers in other parts of the code as well, so I added .0 where it was missing in 7dde6ff6469d2ec5e4b975f9dd0f455987dff313

Reply:
Thanks. If it is in an expression like (a - 1) where a is a float, it's fine because the result will always be a float. In this example, the resulting type depends on if the condition is true or false, this means the code is type unstable and can be a performance issue.

Reply:
ah ok, thanks, good to know!